File: targetFunctions/common/matlabEngineCaller_customLayers/tmwExample/example_loop_single/codegen/lib/matlabEngineCaller_customLayers/matlabEngineCaller_customLayers.c
/*
 * Non-Degree Granting Education License -- for use at non-degree
 * granting, nonprofit, educational organizations only. Not for
 * government, commercial, or other organizational use.
 *
 * matlabEngineCaller_customLayers.c
 *
 * Code generation for function 'matlabEngineCaller_customLayers'
 *
 */

/* Include files */
#include "rt_nonfinite.h"
#include "matlabEngineCaller_customLayers.h"
#include "matlabEngineCaller_customLayers_emxutil.h"
#include "matlabCallFun.h"

/* Function Definitions */
void matlabEngineCaller_customLayers(const double params[8], double contrast,
  const char funcName[10], const char funcPath[122], double bulkIn, double
  bulkOut, emxArray_real_T *output, double sRough_data[], int sRough_size[2])
{
  double outp[300];
  double nLayers;
  double b_params[8];
  char cv0[61];
  int i;
  static const char cv1[17] = { '[', 't', 'o', 't', 'a', 'l', ',', 'l', 'a', 'y',
    'e', 'r', 's', ']', ' ', '=', ' ' };

  char cv2[128];
  static const char cv3[34] = { '(', 'p', 'a', 'r', 'a', 'm', 's', ',', 'b', 'u',
    'l', 'k', '_', 'i', 'n', ',', 'b', 'u', 'l', 'k', '_', 'o', 'u', 't', ',',
    'c', 'o', 'n', 't', 'r', 'a', 's', 't', ')' };

  static const char cv4[4] = { 'c', 'd', ' ', '\'' };

  static const char cv5[2] = { '\'', ';' };

  unsigned int rowCount;
  int colCount;
  int loop_ub;
  double d0;
  sRough_size[0] = 1;
  sRough_size[1] = 1;
  sRough_data[0] = params[0];

  /* Need to reserve some meory for the referencenced variables */
  memcpy(&b_params[0], &params[0], sizeof(double) << 3);
  for (i = 0; i < 17; i++) {
    cv0[i] = cv1[i];
  }

  for (i = 0; i < 10; i++) {
    cv0[i + 17] = funcName[i];
  }

  for (i = 0; i < 34; i++) {
    cv0[i + 27] = cv3[i];
  }

  for (i = 0; i < 4; i++) {
    cv2[i] = cv4[i];
  }

  memcpy(&cv2[4], &funcPath[0], 122U * sizeof(char));
  for (i = 0; i < 2; i++) {
    cv2[i + 126] = cv5[i];
  }

  matlabCallFun(b_params, 8.0, cv0, cv2, bulkIn, bulkOut, contrast, outp,
                &nLayers);
  rowCount = 1U;
  colCount = 0;

  /* reshape the output to [layers * 3] array */
  if (nLayers >= 1.0) {
    i = output->size[0] * output->size[1];
    output->size[0] = (int)nLayers;
    output->size[1] = 3;
    emxEnsureCapacity((emxArray__common *)output, i, (int)sizeof(double));
    loop_ub = (int)nLayers * 3;
    for (i = 0; i < loop_ub; i++) {
      output->data[i] = 0.0;
    }

    d0 = nLayers * 3.0;
    for (i = 0; i < (int)d0; i++) {
      /* Make use of Matlab linear indexing. */
      output->data[((int)rowCount + output->size[0] * colCount) - 1] = outp[i];
      rowCount++;
      if (rowCount == nLayers + 1.0) {
        rowCount = 1U;
        colCount++;
      }
    }
  } else {
    i = output->size[0] * output->size[1];
    output->size[0] = 0;
    output->size[1] = 0;
    emxEnsureCapacity((emxArray__common *)output, i, (int)sizeof(double));
    sRough_size[0] = 0;
    sRough_size[1] = 0;
  }
}

/* End of code generation (matlabEngineCaller_customLayers.c) */

File: Sandbox/matlabEngineCaller/tmwExampleNew/matlabEngineCaller_customLayers.m
function [output,sRough] = matlabEngineCaller_customLayers(params,contrast,funcName,funcPath,bulkIn,bulkOut)


paramsLen = int16(length(params));
% dotM = strfind(funcName,'.m');
% if ~isempty(dotM)
%     funcName = funcName(1:dotM-1);
% end
%[path,funcName,extension] = fileparts(funcName);
funName = ['[total,layers] = ' funcName '(params,bulk_in,bulk_out,contrast)'];
sRough = params(1);

pathCall= ['cd ''' funcPath ''';'];

debug = 0;
if coder.target('MATLAB')
    if debug == 1
        disp('Debug');
        disp(funName);
        disp(pathCall);
        layers= 1;
    end
    dotM = strfind(funcName,'.m');
    if ~isempty(dotM)
        funcName = funcName(1:dotM-1);
    end
    fileHandle = str2func(funcName);
    [output,layers] = fileHandle(params,bulkIn,bulkOut,contrast);
else
    
    path = '/home/arwel/Documents/RascalDev/RAT/Sandbox/matlabEngineCaller/tmwExampleNew';

    
    incPath1 = '/usr/local/MATLAB/R2020a/extern/include';
    %incPath2 = '/usr/include/openmpi-x86_64';
    
    linkPath1 = '/usr/local/MATLAB/R2020a/bin/glnxa64';
    linkFile1 = 'libeng.so';
    linkFile2 = 'libmx.so';
    
    source1 = 'matlabCallFun.c';
    %source2 = 'matlabCallFun.h';
    
    %source1 = 'matlabCallFun.c';
    %source2 = 'matlabCallFun.h';
    
    libPriority = '';
    libPreCompiled = true;
    libLinkOnly = true;

    %coder.cinclude(source2);
    coder.updateBuildInfo('addSourceFiles',source1);
    %coder.updateBuildInfo('addSourceFiles',source2);
    coder.updateBuildInfo('addSourcePaths',path);
    coder.updateBuildInfo('addIncludePaths',incPath1);
    %coder.updateBuildInfo('addIncludePaths',incPath2);
    coder.updateBuildInfo('addLinkObjects',linkFile1,linkPath1,libPriority,libPreCompiled,libLinkOnly);
    coder.updateBuildInfo('addLinkObjects',linkFile2,linkPath1,libPriority,libPreCompiled,libLinkOnly);
    
    %Need to reserve some meory for the referencenced variables
    outp = zeros(1000,3);
    nLayers = 0;
    
    coder.ceval('matlabCallFun', params, paramsLen, funName, pathCall, bulkIn, bulkOut, contrast, coder.wref(outp), coder.wref(nLayers));
    
    rowCount = 1;
    colCount = 1;
    
    m = size(outp,1);
    n = size(outp,3);
    
    %reshape the output to [layers * 3] array
    if (nLayers >= 1)
        output = zeros(nLayers,3);
        for i = 1:(nLayers*3)
            thisVal = outp(i);  %Make use of Matlab linear indexing.
            output(rowCount,colCount) = thisVal;
            rowCount = rowCount + 1;
            if rowCount == (nLayers+1)
                rowCount = 1;
                colCount = colCount + 1;
            end
        end
    else
        output = [1 1 1];
        sRough = 1;
    end

end
